File: app/src/main/java/com/igorganapolsky/vibratingwatchapp/core/util/ViewModelFactory.java
package com.igorganapolsky.vibratingwatchapp.core.util;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.annotation.NonNull;
import com.igorganapolsky.vibratingwatchapp.domain.Repository;
import com.igorganapolsky.vibratingwatchapp.manager.timer.CountdownManager;
import com.igorganapolsky.vibratingwatchapp.presentation.details.TimerDetailsViewModel;
import com.igorganapolsky.vibratingwatchapp.presentation.main.TimerListViewModel;
import com.igorganapolsky.vibratingwatchapp.presentation.settings.SetTimerViewModel;

public class ViewModelFactory implements ViewModelProvider.Factory {

    private static ViewModelFactory factory;

    private final Repository repository;
    private final CountdownManager countdownManager;

    private ViewModelFactory(Repository repository, CountdownManager countdownManager) {
        this.repository = repository;
        this.countdownManager = countdownManager;
    }

    public static void initFactory(Repository repository,
                                   CountdownManager countdownManager) {
        if (factory == null) {
            factory = new ViewModelFactory(repository, countdownManager);
        }
    }

    public static ViewModelProvider.Factory getInstance() {
        return factory;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        ViewModel viewModel;
        if (modelClass.isAssignableFrom(SetTimerViewModel.class)) {
            viewModel = new SetTimerViewModel(repository);
        } else if (modelClass.isAssignableFrom(TimerListViewModel.class)) {
            viewModel = new TimerListViewModel(repository, countdownManager);
        } else if (modelClass.isAssignableFrom(TimerDetailsViewModel.class)) {
            viewModel = new TimerDetailsViewModel(repository, countdownManager);
        } else {
            throw new IllegalStateException("No associated view model with " + modelClass.getName());
        }
        return (T) viewModel;
    }
}

File: app/src/main/kotlin/com/igorganapolsky/vibratingwatchapp/common/ViewModelFactory.kt
package com.igorganapolsky.vibratingwatchapp.common

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.igorganapolsky.vibratingwatchapp.data.ITimersRepository
import com.igorganapolsky.vibratingwatchapp.presentation.new_timer_creation_screen.viewmodel.NewTimerViewModel
import com.igorganapolsky.vibratingwatchapp.presentation.home_screen.viewmodel.TimersListViewModel
import com.igorganapolsky.vibratingwatchapp.presentation.timer_edit_screen.viewmodel.ExistingTimerViewModel

class ViewModelFactory private constructor(
    private val repository: ITimersRepository,
    private val countdownManager: ICountdownController
) : ViewModelProvider.Factory {

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        val viewModel: ViewModel
        if (modelClass.isAssignableFrom(NewTimerViewModel::class.java)) {
            viewModel =
                NewTimerViewModel(
                    repository
                )
        } else if (modelClass.isAssignableFrom(TimersListViewModel::class.java)) {
            viewModel =
                TimersListViewModel(
                    repository,
                    countdownManager
                )
        } else if (modelClass.isAssignableFrom(ExistingTimerViewModel::class.java)) {
            viewModel =
                ExistingTimerViewModel(
                    repository,
                    countdownManager
                )
        } else {
            throw IllegalStateException("No associated view model with " + modelClass.name)
        }
        return viewModel as T
    }

    companion object {
        private lateinit var factory: ViewModelFactory

        fun initFactory(repository: ITimersRepository, countdownController: ICountdownController) {
            factory =
                ViewModelFactory(
                    repository,
                    countdownController
                )
        }

        val instance: ViewModelProvider.Factory?
            get() = factory
    }
}
